### PR TITLE
chore(core): Migrate centreon-web to TypeScript

### PR DESCRIPTION
**Pull request by Thebarda**
# Pull Request Template

## Description

This PR migrate each Javascript file to Typescript for improve typing.
It also integrate Typescript into unit tests, parser (babel), linter (eslint) and build tool (webpack).

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [ ] 19.10.x
- [x] 20.04.x (master)

<h2> How this pull request can be tested ? </h2>

Build the project then launch it.
- Go to Status details > services and navigate through the breadcrumb.
- Poller menu, host menu, service menu and user menu should work.
- Go to pollers list (via poller menu), click on 'Add server with wizard' button then add a Centreon Remote Server
- Go to pollers list (via poller menu), click on 'Add server with wizard' button then add a Centreon Poller

#### Community contributors & Centreon team

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [x] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1 +1,20 @@
...
+        map: [
+          ['@centreon/ui', '@centreon/ui/src'],
+        ],
+        extensions: ['.js', '.jsx',]
+      }
+    }
+  },
+  rules: {
+    "react/jsx-filename-extension": [2, { "extensions": [".jsx", ".tsx"] }]
+  }
```
Comment (bdauria) [inline .eslintrc.js]: can we move these rules into eslint-config-centreon?

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,65 @@
+{
+  "compilerOptions": {
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    // "lib": [],                             /* Specify library files to be included in the compilation. */
```
Comment (bdauria) [inline tsconfig.json]: I think we can remove the comments and rely on the actual documentation

File: www/front_src/src/App.tsx [hunk 3]
```diff
@@ -11,40 +11,47 @@ import { connect } from 'react-redux';
 import { ConnectedRouter } from 'connected-react-router';
 import Fullscreen from 'react-fullscreen-crossbrowser';
 import queryString from 'query-string';
-import { batchActions } from 'redux-batched-actions';
 import { StylesProvider } from '@material-ui/styles';
-import Header from './components/header';
-import { history } from './store';
+import Header from './components/header/index.tsx';
```
Comment (bdauria) [inline www/front_src/src/App.tsx]: we shouldn't have to specify index files specifically

File: www/front_src/src/App.tsx [hunk 4]
```diff
@@ -11,40 +11,47 @@ import { connect } from 'react-redux';
...
-import MainRouter from './components/mainRouter';
-import axios from './axios';
+import NavigationComponent from './components/navigation/index.tsx';
+import Tooltip from './components/tooltip/index.tsx';
+import Footer from './components/footer/index.tsx';
+import MainRouter from './components/mainRouter/index.tsx';
+import axios from './axios/index.ts';
 
-import { fetchExternalComponents } from './redux/actions/externalComponentsActions';
+import { fetchExternalComponents } from './redux/actions/externalComponentsActions.ts';
```
Comment (bdauria) [inline www/front_src/src/App.tsx]: we shouldn't have to specify extensions specifically

File: www/front_src/src/App.tsx [hunk 5]
```diff
@@ -11,40 +11,47 @@ import { connect } from 'react-redux';
...
 
 import styles from './App.scss';
 import footerStyles from './components/footer/footer.scss';
 import contentStyles from './styles/partials/_content.scss';
 
-class App extends Component {
-  state = {
+interface Props {
+  fetchExternalsComponents: Function;
```
Comment (bdauria) [inline www/front_src/src/App.tsx]: Doesn't ESLint complain about this? I guess it should be () => undefined

File: www/front_src/src/components/navigation/index.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import reactRoutesSelector from '../../redux/selectors/navigation/reactRoutes.ts';
+import { fetchNavigationData } from '../../redux/actions/navigationActions.ts';
+
+interface Props {
+  fetchNavigationData: Function;
+  navigationData: Array;
+  reactRoutes: object;
+}
+
+class Navigation extends Component<PRops> {
```
Comment (bdauria) [inline www/front_src/src/components/navigation/index.tsx]: There is a typo, it should be `Props`

File: www/front_src/src/redux/actions/bamConfigurationActions.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+export const SET_BA_CONFIGURATION = '@bam/SET_BA_CONFIGURATION';
```
Comment (bdauria) [inline www/front_src/src/redux/actions/bamConfigurationActions.ts]: This file is not needed anymore and should be removed

File: www/front_src/src/redux/reducers/bamConfigurationReducer.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
+/* eslint-disable camelcase */
```
Comment (bdauria) [inline www/front_src/src/redux/reducers/bamConfigurationReducer.ts]: This file is not needed anymore and should be removed

File: .eslintrc.js (hunk 1 above)
Comment (Thebarda) [reply .eslintrc.js]: Of course

File: www/front_src/src/App.tsx (hunk 3 above)
Comment (Thebarda) [reply www/front_src/src/App.tsx]: Linter complains about it...

File: www/front_src/src/App.tsx (hunk 4 above)
Comment (Thebarda) [reply www/front_src/src/App.tsx]: Same here linter complains about it with...
`Missing file extension for`

**Comment by Thebarda** [comment]
With @bdauria, we have decided to close this PR and create a new one with a limited scope to `App.tsx` and tool config (webpack, eslint, babel and jest). This decision will avoid to have a lot of files modified.
It also follows our logic to migrate centreon-web to typescript step by step.

**Pull request closed without merging**